File: web/translation.js
import {
  englishPortuguese,
  maxTranslationWords,
} from "./engPorDictionary.js";
import { tatoebaEnglishPortuguese } from "./tatoebaPhrases.js";

const wordPattern = /[A-Za-z]+(?:[-'][A-Za-z]+)*/gu;

export function translateEnglishToPortuguese(value) {
  const exact = translateExactSentence(value);
  if (exact !== "") return exact;

  const matches = Array.from(value.matchAll(wordPattern));
  if (matches.length === 0) return value;

  const translated = [];
  let cursor = 0;
  let index = 0;
  while (index < matches.length) {
    const match = findLongestMatch(value, matches, index);
    if (!match) {
      const word = matches[index];
      translated.push(value.slice(cursor, word.index));
      translated.push(word[0]);
      cursor = word.index + word[0].length;
      index += 1;
      continue;
    }

    translated.push(value.slice(cursor, match.start));
    translated.push(matchCase(value.slice(match.start, match.end), match.replacement));
    cursor = match.end;
    index = match.nextIndex;
  }

  translated.push(value.slice(cursor));
  return translated.join("");
}

function translateExactSentence(value) {
  const translated = tatoebaEnglishPortuguese[normalizeKey(value)];
  if (translated) return matchCase(value, translated);

  const [, phrase = value, punctuation = ""] = value.trim().match(/^(.*?)([.!?]+)?$/u);
  const withoutPunctuation = tatoebaEnglishPortuguese[normalizeKey(phrase)];
  return withoutPunctuation ? matchCase(phrase, withoutPunctuation) + punctuation : "";
}

function findLongestMatch(value, matches, startIndex) {
  const limit = Math.min(matches.length, startIndex + maxTranslationWords);
  for (let endIndex = limit; endIndex > startIndex; endIndex -= 1) {
    const start = matches[startIndex].index;
    const last = matches[endIndex - 1];
    const end = last.index + last[0].length;
    const candidate = value.slice(start, end);
    if (!containsOnlyWordSeparators(candidate)) continue;

    const translations = englishPortuguese[normalizeKey(candidate)];
    if (translations?.length) {
      return { start, end, replacement: translations[0], nextIndex: endIndex };
    }
  }
  return null;
}

function containsOnlyWordSeparators(value) {
  return /^[A-Za-z]+(?:[-' ]+[A-Za-z]+)*$/u.test(value);
}

function normalizeKey(value) {
  return value.toLowerCase().replace(/\s+/gu, " ").trim();
}

function matchCase(source, translated) {
  const words = Array.from(source.matchAll(wordPattern), (match) => match[0]);
  if (words.length > 0 && words.every((word) => word.toUpperCase() === word)) {
    return translated.toUpperCase();
  }
  if (/^[A-Z]/u.test(source)) {
    return translated.charAt(0).toUpperCase() + translated.slice(1);
  }
  return translated;
}

File: src/geobe/translation.py
"""English-to-Portuguese translation helpers backed by FreeDict data."""

from __future__ import annotations

import re

from geobe.eng_por_dictionary import ENGLISH_PORTUGUESE, MAX_TRANSLATION_WORDS
from geobe.tatoeba_phrases import TATOEBA_ENGLISH_PORTUGUESE

WORD_RE = re.compile(r"[A-Za-z]+(?:[-'][A-Za-z]+)*")


def translate_english_to_portuguese(text: str) -> str:
    """Translate English words and phrases to Portuguese dictionary equivalents."""
    phrase = _translate_exact_sentence(text)
    if phrase is not None:
        return phrase

    matches = list(WORD_RE.finditer(text))
    if not matches:
        return text

    translated: list[str] = []
    cursor = 0
    index = 0
    while index < len(matches):
        match = _find_longest_match(text, matches, index)
        if match is None:
            word = matches[index]
            translated.append(text[cursor : word.start()])
            translated.append(word.group(0))
            cursor = word.end()
            index += 1
            continue

        start, end, replacement, next_index = match
        translated.append(text[cursor:start])
        translated.append(_match_case(text[start:end], replacement))
        cursor = end
        index = next_index

    translated.append(text[cursor:])
    return "".join(translated)


def _translate_exact_sentence(text: str) -> str | None:
    translated = TATOEBA_ENGLISH_PORTUGUESE.get(_normalize_key(text))
    if translated is not None:
        return _match_case(text, translated)
    phrase, punctuation = _split_terminal_punctuation(text)
    translated = TATOEBA_ENGLISH_PORTUGUESE.get(_normalize_key(phrase))
    if translated is None:
        return None
    return _match_case(phrase, translated) + punctuation


def _find_longest_match(
    text: str,
    matches: list[re.Match[str]],
    start_index: int,
) -> tuple[int, int, str, int] | None:
    limit = min(len(matches), start_index + MAX_TRANSLATION_WORDS)
    for end_index in range(limit, start_index, -1):
        start = matches[start_index].start()
        end = matches[end_index - 1].end()
        candidate = text[start:end]
        if not _contains_only_word_separators(candidate):
            continue

        translations = ENGLISH_PORTUGUESE.get(_normalize_key(candidate))
        if translations:
            return start, end, translations[0], end_index
    return None


def _contains_only_word_separators(value: str) -> bool:
    return re.fullmatch(r"[A-Za-z]+(?:[-' ]+[A-Za-z]+)*", value) is not None


def _normalize_key(value: str) -> str:
    return re.sub(r"\s+", " ", value.lower()).strip()


def _split_terminal_punctuation(text: str) -> tuple[str, str]:
    match = re.search(r"([.!?]+)$", text.strip())
    if match is None:
        return text, ""
    return text[: match.start()], match.group(1)


def _match_case(source: str, translated: str) -> str:
    words = WORD_RE.findall(source)
    if words and all(word.isupper() for word in words):
        return translated.upper()
    if source[:1].isupper():
        return translated[:1].upper() + translated[1:]
    return translated
